From 0c05d99d67f083d61fef5b40c1663773e46a14fc Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Wed, 11 Jan 2023 09:28:04 +0100
Subject: [PATCH 0255/1695] drivers: stm32_risaf: enforce encryption control on
 configuration

The encryption is only available on secure area. Other configuration
must be removed.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git a7aea82c7ee49c5e1b6a2883d64ba7b7886ba3a8 drivers: stm32_risaf: enforce encryption control on configuration]
Change-Id: I421ab1bd1a0b69eb6255b8ac7d819793adfbc92f
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
---
 core/drivers/firewall/stm32_risaf.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/core/drivers/firewall/stm32_risaf.c b/core/drivers/firewall/stm32_risaf.c
index d8b65a519..2908b21d3 100644
--- a/core/drivers/firewall/stm32_risaf.c
+++ b/core/drivers/firewall/stm32_risaf.c
@@ -238,10 +238,18 @@ static TEE_Result risaf_configure_region(struct stm32_risaf_instance *risaf,
 			_RISAF_REG_CFGR_ALL_MASK,
 			cfg & _RISAF_REG_CFGR_ALL_MASK);
 
-	if (((cfg & _RISAF_REG_CFGR_ENC) == _RISAF_REG_CFGR_ENC) &&
-	    !risaf->pdata.enc_supported) {
-		EMSG("RISAF %#"PRIxPTR": encryption feature error", risaf->pdata.base.pa);
-		return -1;
+	if ((cfg & _RISAF_REG_CFGR_ENC) == _RISAF_REG_CFGR_ENC) {
+		if (!risaf->pdata.enc_supported) {
+			EMSG("RISAF %#"PRIxPTR": encryption feature error",
+			     risaf->pdata.base.pa);
+			return TEE_ERROR_GENERIC;
+		}
+
+		if ((cfg & _RISAF_REG_CFGR_SEC) != _RISAF_REG_CFGR_SEC) {
+			EMSG("RISAF %#"PRIxPTR": encryption on non secure area",
+			     risaf->pdata.base.pa);
+			return TEE_ERROR_GENERIC;
+		}
 	}
 
 	DMSG("RISAF %#"PRIxPTR": region %02d - start 0x%08x - end 0x%08x - cfg 0x%08x - cidcfg 0x%08x",
-- 
2.47.2

